src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt

```diff
@@ -1,4 +1,49 @@
 #nullable enable
+OpenTelemetry.Logs.LogEmitter
+OpenTelemetry.Logs.LogEmitter.Log(in OpenTelemetry.Logs.LogRecordData data, in OpenTelemetry.Logs.LogRecordAttributeList attributes = default(OpenTelemetry.Logs.LogRecordAttributeList)) -> void
+OpenTelemetry.Logs.LogEmitter.LogEmitter(OpenTelemetry.Logs.OpenTelemetryLoggerProvider! loggerProvider) -> void
 OpenTelemetry.Logs.LogRecord.FormattedMessage.set -> void
+OpenTelemetry.Logs.LogRecord.GetDataRef() -> OpenTelemetry.Logs.LogRecordData
 OpenTelemetry.Logs.LogRecord.State.set -> void
 OpenTelemetry.Logs.LogRecord.StateValues.set -> void
+OpenTelemetry.Logs.LogRecordAttributeList
```

Comment:
Since .NET uses the term `Tag` for metrics and traces, should we stay consistent and call 'em tags for logs as well?

Reply:
I struggled with that one. We have `State` & `StateValues` on `LogRecord` today. The most consistent thing might be to name this `LogRecordStateValueList` and not introduce another concept? I ended up going with "attributes" because that is what the spec calls them and the `LogEmitter` API right now is pretty close to spec but open to whatever.

Reply:
🤔 yea I hear ya on the struggle. I think sticking to `Attribute` to align with the spec might make sense in this context. Though also maybe this is a good enough reason to make `LogRecordAttributeList` internal for now. I guess I'm not too clear yet how all this may play out in the context of a logging framework besides ILogger.

Reply:
Pretty much, just ignore `ILogger` 😄 What `LogEmitter` does is take a `LogRecord` from the pool, applies the `LogRecordData` + `LogRecordAttributeList` to it, and then sends it at the `LogProcessor` at the root of the `OpenTelemetryLoggerProvider`. It doesn't really flow through `ILogger` at all.

I haven't tested this yet, but to help illustrate here is the Serilog sink I scratched out to mess with the `LogEmitter` API:

```csharp
    public static class OpenTelemetrySerilogSinkExtensions
    {
        public static LoggerConfiguration OpenTelemetry(
            this LoggerSinkConfiguration loggerConfiguration,
            OpenTelemetryLoggerProvider openTelemetryLoggerProvider)
        {
            Guard.ThrowIfNull(loggerConfiguration);

            return loggerConfiguration.Sink(new OpenTelemetrySerilogSink(openTelemetryLoggerProvider));
        }
    }

    internal sealed class OpenTelemetrySerilogSink : ILogEventSink
    {
        private readonly LogEmitter logEmitter;

        public OpenTelemetrySerilogSink(OpenTelemetryLoggerProvider openTelemetryLoggerProvider)
        {
            this.logEmitter = new LogEmitter(openTelemetryLoggerProvider);
        }

        public void Emit(LogEvent logEvent)
        {
            Debug.Assert(logEvent != null, "LogEvent was null.");

            LogRecordData data = new(Activity.Current)
            {
                Timestamp = logEvent!.Timestamp.UtcDateTime,
                LogLevel = (LogLevel)(int)logEvent.Level,
                Message = logEvent.MessageTemplate.Text,
                Exception = logEvent.Exception,
            };

            LogRecordAttributeList attributes = default;
            foreach (KeyValuePair<string, LogEventPropertyValue> property in logEvent.Properties)
            {
                // TODO: Serilog supports complex type logging. This is not yet
                // supported in OpenTelemetry.
                if (property.Key == Constants.SourceContextPropertyName)
                {
                    data.CategoryName = property.Value.ToString();
                }
                else if (property.Value is ScalarValue scalarValue)
                {
                    attributes.Add(property.Key, scalarValue.Value);
                }
                else if (property.Value is SequenceValue sequenceValue)
                {
                    IReadOnlyList<LogEventPropertyValue> elements = sequenceValue.Elements;
                    if (elements.Count > 0 && elements[0] is ScalarValue)
                    {
                        object[] values = new object[elements.Count];
                        for (int i = 0; i < elements.Count; i++)
                        {
                            if (elements[i] is ScalarValue value)
                            {
                                values[i] = value.Value;
                            }
                        }

                        attributes.Add(property.Key, values);
                    }
                }
            }

            this.logEmitter.Log(in data, in attributes);
        }
    }
```

Reply:
Cool, yea this example is great. Is the reason due to size/performance that `LogRecordData` doesn't have an LogRecordAttributeList field? ... Oh, or maybe the reason is that if you were to do `this.logEmitter.Log(in data)` that the `in data` parameter would not preclude the attributes from being modified.

Reply:
Cool, yea this example is great. Is the reason due to size/performance that `LogRecordData` doesn't have an LogRecordAttributeList field? ... Oh, or maybe the reason is that if you were to do `this.logEmitter.Log(in data)` that the `in data` parameter would not preclude the attributes from being modified.

Reply:
Originally I had the attributes as a field on `LogRecordData` but I ran into some weird compiler limitations. What I wanted to do was allow people to do this...

```csharp
logEmitter.Log(new()
{
   Message = "Hello world",
   Attributes = { ["count" ] = 1, ["id"] = 18 }
})
```

That would work if "Attributes" was a reference type, but compiler doesn't allow with a value type! I tried a bunch of things, ultimately moved it to a separate parameter. At the moment you can do...

```csharp
logEmitter.Log(
   new() { Message = "Hello world" },
   new() { ["count" ] = 1, ["id"] = 18 }
})
```

Besides that...

* Having it separate keeps the size of `LogRecordData` down. Now `LogRecord`s live in a pool and store a `LogRecordData`. Having the "attributes" as part of `LogRecordData` would mean that pool would take up a bunch more memory.

* I could see people mess it up. Imagine something like this...

   ```csharp
   LogRecord data = new()
   {
      Message = "Hello world"
   };
   
   BuildAttributes(data.Attributes); // This passes a copy. So anything added in the method won't really be added.
   
   logEmitter.Log(in data);
   ```

* Having it separate leaves some room for expansion. What I'm working towards here is potentially using `LogEmitter` to support logging of complex types. Maybe the API will end up like this:

   ```csharp
   LogEmitter.Log(in LogRecordData data, in LogRecordAttributeList attributes = default, in LogRecordObjectWriter objectWriter = default)
   ```

   Basically you always pass in the core sutff, but then you can pass in other things that depend on the case. I have no idea what it will ultimately look like but something I'm thinking about.

At one point I tried a builder pattern like...

```csharp
logEmitter.Log(new()
   .SetMessage("Hello world")
   .AddAttribute("count", 1))
```

But that is tricky to do with a struct. Each time you return the struct it is a new copy so you have to write it like ```public static ref LogRecordBuilder SetMessage(this ref LogRecordBuilder builder)```. Seemed gross 🤣